### ISSUE TITLE
Fatal Error : Class Not found phpsmug\Client

### ISSUE THREAD

**Issue by davinci-007** [lildude/phpSmug#39]
Hello there,
I intalled compose using SSH2 and it has got installed in ~/.composer directory. I have uploaded to the zip files to ~/smugmug/php and library folder at ~/smugmug/php/lib/phpSmug where Client.php resides.
Also note that the vendor directory is installed in ~/.composer/vendor with autoloader.php

I ran into 2 issues 
1) When running composer update , it throws "Problem 1
    - The requested package lildude/phpsmug No version set (parsed as 1.0.0) is satisfiable by lildude/phpsmug[No version set (parsed as 1.0.0)] but these conflict with your requirements or minimum-stability". Well, my composer.json is already having this
 **"require": {
        "lildude/phpsmug": "^4.0",
        "php": ">=5.5.0",
        "guzzlehttp/guzzle": "~6.1",
        "guzzlehttp/oauth-subscriber": "0.3.*"
    },**

2) Ignoring this, I am unable to load phpSmug\Client class and shows this error **Fatal error: Class 'phpSmug\Client' not found in /home2/visionone/public_html/smugmug/php/example.php on line 49**

**Comment by lildude** [comment]
Without seeing the full structure of your application or knowing precisely what you've done, it's a bit hard to tell why you're encountering these problems but I suspect your problems may be around your composer usage.

Are you able to successfully install phpSmug independently of your application using: `composer require lildude/phpsmug`?

As an example, here's a very quick run through I've just done:

```
$ mkdir foobar
$ cd foobar
$ composer require lildude/phpsmug
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Using version ^4.0 for lildude/phpsmug
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing guzzlehttp/promises (v1.3.1)
    Loading from cache

  - Installing psr/http-message (1.0.1)
    Loading from cache

  - Installing guzzlehttp/psr7 (1.4.2)
    Loading from cache

  - Installing guzzlehttp/guzzle (6.3.0)
    Downloading: 100%

  - Installing guzzlehttp/oauth-subscriber (0.3.0)
    Loading from cache

  - Installing lildude/phpsmug (4.0.0)
    Loading from cache

guzzlehttp/guzzle suggests installing psr/log (Required for using the Log middleware)
Writing lock file
Generating autoload files
$
$ ls
composer.json composer.lock vendor
$ ls vendor
autoload.php composer     guzzlehttp   lildude      psr
$ ls vendor/lildude
phpsmug
$
```

At this point, I can create a simple `index.php` containing:

```
<?php
require_once 'vendor/autoload.php';
$client = new phpSmug\Client('bogusapikey');
print_r($client);
?>
```

... and running this works a treat:

```
$ php index.php
phpSmug\Client Object
(
    [AppName] => Unknown Application
    [APIKey] => bogusapikey
    [OAuthSecret] =>
    [_verbosity] =>
    [_shorturis] =>
    [oauth_token] =>
    [oauth_token_secret] =>
    [stack:phpSmug\Client:private] =>
    [client:phpSmug\Client:private] =>
    [httpClient:phpSmug\Client:private] => GuzzleHttp\Client Object
        (
            [config:GuzzleHttp\Client:private] => Array
                (
                    [base_uri] => GuzzleHttp\Psr7\Uri Object
[... truncated for brevity ...]
``

If this works, it means there's probably a problem with your `composer.json` file (problem 1) and you're probably referencing the wrong directory in your application when calling the `autoload.php` which is causing problem 2.

**Comment by davinci-007** [comment]
1. My current PWD is ~/.composer 
2. and I just gave the command  **"composer require lildude/phpsmug"** which gave me this 

Using version ^4.0 for lildude/phpsmug
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package lildude/phpsmug No version set (parsed as 1.0.0) is satisfiable by lildude/phpsmug[No version set (parsed as 1.0.0)] but these conflict with your requirements or minimum-stability.
Installation failed, reverting ./composer.json to its original content.

my copy of composer.json which is residing in ~/.composer is as follows. The one in bold is what i have changed

{

    "name": "lildude/phpsmug",
    "type": "library",
    "description": "A simple object orientated wrapper for the SmugMug API v2",
    "homepage": "http://phpsmug.com",
    "keywords": ["smugmug", "api"],
    "license": "MIT",
    "authors": [
        {
            "name": "Colin Seymour",
            "homepage": "http://lildude.co.uk"
        },
        {
            "name": "phpSmug Contributors",
            "homepage": "https://github.com/lildude/phpSmug/graphs/contributors"
        }
    ],
    "require": {
        **"lildude/phpsmug": "^4.0",**
        "php": ">=5.5.0",
        "guzzlehttp/guzzle": "~6.1",
        "guzzlehttp/oauth-subscriber": "0.3.*"
    },
    "require-dev": {
        "phpunit/phpunit": "~4.0",
        "fabpot/php-cs-fixer": "~1.10",
        "satooshi/php-coveralls": "dev-master"
    },
   "autoload": {
    "psr-4": {
      **"phpSmug\\":"/home2/visionone/smugmug/php/lib/phpSmug/"**
    }
  },
    "extra": {
        "branch-alias": {
            "dev-master": "4.0.x-dev"
        }
    }
}

**Comment by lildude** [comment]
> My current PWD is ~/.composer

You've not done as I've requested. Please attempt in a completely clean and empty directory. 

As I mentioned before, it appears your problem may be related to your current `composer.json` or your usage. By testing in a completely new and empty directory you can confirm that things are indeed working as expected _without_ your `composer.json` file. You can then compare the automatically generated `composer.json` file to the one you're having problems with and then make the necessary adjustments.

My example usage confirms there is _no_ problem with phpSmug.

**Comment by davinci-007** [comment]
Great dude, it worked like a charm ! but now when i run 1st file example.php, can you tell me whats wrong am i doing? I havent edited example.php, I just pasted my username and my Key.  When I load example.php, its get stuck with this error
**"Warning: Invalid argument supplied for foreach()"** at line 53 which is 

foreach ($images->AlbumImage as $image) {

**Comment by lildude** [comment]
Great!!

As for your new error, did you change these variables at the top...

https://github.com/lildude/phpSmug/blob/d6ddb55337f17097f2d03d9e8745f78963b96dff/examples/example.php#L19-L21

... to match your credentials etc?

**Comment by davinci-007** [comment]
yes sir, I did !

**Comment by lildude** [comment]
:+1: in that case, the problem is going to be the first public album found probably doesn't have any images in it or if it does, they may not be public or accessible to the user without authentication.

You can confirm this by adding `print_r($images)` just after this line:

https://github.com/lildude/phpSmug/blob/d6ddb55337f17097f2d03d9e8745f78963b96dff/examples/example.php#L51

**Comment by davinci-007** [comment]
Yes, I only have private albums, in this case how will those get displayed?

**Comment by lildude** [comment]
You'll need to authenticate. https://github.com/lildude/phpSmug/blob/master/examples/example-oauth.php is the example you want.

**Comment by davinci-007** [comment]
yes it's working partly,  it comes back to my callback url but I need to fetch images from specific album so where do I need to specify the uri/album key in example_oath file?

**Comment by lildude** [comment]
If you know the exact album, you can change `$albums->Album[0]->Uris->AlbumImages` at https://github.com/lildude/phpSmug/blob/d6ddb55337f17097f2d03d9e8745f78963b96dff/examples/example-oauth.php#L80 to the exact path as per the API.

For example, if the album ID is `123456`, you'd change that line to:

```php
$images = $client->get('album/123456!images');
```

... or determine it by reviewing the results of the earlier call to get all the albums:

https://github.com/lildude/phpSmug/blob/d6ddb55337f17097f2d03d9e8745f78963b96dff/examples/example-oauth.php#L78

Removing the `, array('count' => 1)` part will return all albums.

I highly recommend you to play with the examples by putting in `print_r()` statements so you can see what each API call in the examples do and return.  Play with [SmugMug's live API browser](https://api.smugmug.com/api/v2) too.

**Comment by davinci-007** [comment]
Thanks for that tip, but as i said earlier, when i use example-oauth.php or example-external-links.php, both redirect me to smugmug login / verify login and once i reinput my credentials, there comes a 2 sec splash page for App Authorize and then it lands me back to my callback url, and the script runs further.

I did also comment the "Authorize" anchor echo tag, and uncommented the header location line, but it still does no good.

How do i keep my authorization alive and even if i do page refresh, it should just run the script rather than redirecting me?

**Comment by lildude** [comment]
> I did also comment the "Authorize" anchor echo tag, and uncommented the header location line, but it still does no good.

Yes, this is expected as the example is very simple and does not store the returned authentication token for security reasons.

> How do i keep my authorization alive and even if i do page refresh, it should just run the script rather than redirecting me?

You'll need to save the authentication token returned by the API to some location of your choice and then use it as detailed in the docs at https://github.com/lildude/phpSmug/blob/master/docs/authentication.md
